x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.sql.plugin;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.xpack.sql.action.SqlQueryRequest;
+import org.elasticsearch.xpack.sql.proto.Mode;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.sql.proto.Protocol.URL_PARAM_FORMAT;
+
+public class SqlMediaTypeParser {
+
+    static class SqlMediaType {
+        private final XContentType xContentType;
+        private final TextFormat textFormat;
+        private final boolean isTextFormat;
+
+        private SqlMediaType(XContentType xContentType, TextFormat textFormat) {
+            this.xContentType = xContentType;
+            this.textFormat = textFormat;
+            isTextFormat = textFormat != null;
+        }
+
+        private static SqlMediaType xContentType(XContentType xContentType) {
+            return xContentType != null ? new SqlMediaType(xContentType, null) : null;
+        }
+
+        private static SqlMediaType textFormat(TextFormat textFormat) {
+            return textFormat != null ? new SqlMediaType(null, textFormat) : null;
+        }
+
+        public boolean isTextFormat() {
+            return isTextFormat;
+        }
+
+        public XContentType xContentType() {
+            return xContentType;
+        }
+
+        public TextFormat textFormat() {
+            return textFormat;
+        }
+
+        public static SqlMediaType fromMediaTypeOrFormat(String mediaType) {
+            XContentType xContentType = XContentType.fromMediaTypeOrFormat(mediaType);
+            return xContentType != null ? xContentType(xContentType) : textFormat(TextFormat.fromMediaTypeOrFormat(mediaType));
+        }
+    }
+    /*
+     * Since we support {@link TextFormat} <strong>and</strong>
+     * {@link XContent} outputs we can't use {@link RestToXContentListener}
+     * like everything else. We want to stick as closely as possible to
+     * Elasticsearch's defaults though, while still layering in ways to
+     * control the output more easily.
+     *
+     * First we find the string that the user used to specify the response
+     * format. If there is a {@code format} parameter we use that. If there
+     * isn't but there is a {@code Accept} header then we use that. If there
+     * isn't then we use the {@code Content-Type} header which is required.
+     */
+    public static SqlMediaType getResponseMediaType(RestRequest request, SqlQueryRequest sqlRequest) {
```

Comment:
Do those methods need to be public? As far as I can see they're used only for testing, could they be package private?

Reply:
I was also wondering: which methods need to be public for the backport?

Reply:
Good point. Accessibility corrected.